### PR TITLE
thunder.jit: add post_optimization_transform

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes: #533

### CODE REVIEW

**Comment by t-vi** [comment]
When we make this switch, let us use a different method than `__call__` on these, e.g. `transform_traces` or so.
Then we could have the transforming of the modules (for early transforms) in `transform_module`, and #483 can add it's methods.

File: thunder/__init__.py [hunk 1]
```diff
@@ -1,5 +1,5 @@
 from functools import wraps
-from typing import Any
+from typing import Any, List
```
Comment (crcrpar) [inline thunder/__init__.py:2]:
```suggestion
from typing import Any
```

**Review by crcrpar — COMMENTED** [review]
this covers what I have in my mind. great

File: thunder/__init__.py [hunk 2]
```diff
@@ -592,6 +599,17 @@ def get_computation_and_inputs(*args, **kwargs):
 
             if not compile_options.get("disable_inplace_copy_check", False):
                 thunder.core.transform_common._inplace_copy_sanity_check(computation_trc)
+
+            for transform in post_optimization_transforms:
+                # NOTE: `backward_trc` could be None.
```
[the remark this answers is not in the document]
Comment (kshitij12345) [reply thunder/__init__.py:604]: I think this is fine, example - I have some transformation for profiling, it is ok to not have `backward_trc` as I maybe only interested in profiling the forward trace for now.

File: thunder/tests/test_jit_general.py [hunk 3]
```diff
@@ -813,3 +813,45 @@ def foo(a, scalar):
...
+                        bsyms.append(comment_bsym)
+
+                    bsyms.append(bsym)
+
+                commented_trace.bound_symbols = bsyms
+                return commented_trace
+
+            return _comment_add_op_transform(fwd_trc), (
+                _comment_add_op_transform(bwd_trc) if bwd_trc is not None else None
+            )
```
Comment (IvanYashchuk) [inline thunder/tests/test_jit_general.py:842]: I think the transform should take just one trace and "if backward is not None" should be in infra code not in transform code.

Comment (kshitij12345) [reply thunder/tests/test_jit_general.py:842]: Makes sense, also linking the relevant [comment](https://github.com/Lightning-AI/lightning-thunder/issues/533#issuecomment-2152894973) from the issue here

**Pull request merged**